sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala

```diff
@@ -27,15 +27,20 @@ abstract class FileScanBuilder(
     dataSchema: StructType) extends ScanBuilder with SupportsPushDownRequiredColumns {
   private val partitionSchema = fileIndex.partitionSchema
   private val isCaseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
+  protected val supportsNestedSchemaPruning: Boolean = false
   protected var requiredSchema = StructType(dataSchema.fields ++ partitionSchema.fields)
 
   override def pruneColumns(requiredSchema: StructType): Unit = {
+    // [SPARK-30107] While the passed `requiredSchema` always have pruned nested columns, the actual
+    // data schema of this scan is determined in `readDataSchema`. File formats that don't support
+    // nested schema pruning, use `requiredSchema` as a reference and perform the pruning partially.
     this.requiredSchema = requiredSchema
   }
 
   protected def readDataSchema(): StructType = {
     val requiredNameSet = createRequiredNameSet()
-    val fields = dataSchema.fields.filter { field =>
+    val schema = if (supportsNestedSchemaPruning) requiredSchema else dataSchema
```

Comment:
If supportsNestedSchemaPruning is true but SQLConf.get.nestedSchemaPruningEnabled is disabled? We should use `dataSchema`?

Reply:
If `SQLConf.get.nestedSchemaPruningEnabled` is false, we will use the old path without nested schema pruning. So, the passed `requiredSchema` will always be the source of truth for ORC and Parquet. I've added a test to confirm this to `SchemaPruningSuite`.

Reply:
hmm, so what is the difference using `requiredSchema` or `dataSchema`? When using the old path without nested schema pruning, the passed `requiredSchema` is always the source of truth. When `supportsNestedSchemaPruning` is false, we turn to `dataSchema`??

Reply:
The difference between those two is that `requiredSchema` can have pruned nested columns if nested schema pruning is enabled. Therefore, ORC and Parquet must always use data types from `requiredSchema`. If nested schema pruning is disabled, `requiredSchema` will have pruned only top-level columns. In that case, it is still safe to use data types from `requiredSchema`. CSV/JSON/Avro will always use `requiredSchema` as a reference but keep data types from `dataSchema`, so that even if Spark passes a schema with pruned nested columns, we still use data types defined in `dataSchema` to match the behavior before this PR.